[Feature][PostgreSQL] Compile expression functions

rldhont opened the pull request.

rldhont committed:
[Feature][PostgreSQL] Compile expression functions

rldhont commented:
@mdouchin it's a work in progress

nyalldawson commented on src/providers/postgres/qgspostgresexpressioncompiler.cpp:
```diff
@@ -32,3 +36,91 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
+      }
+      else
+      {
+        nd = sqlFunctionFromFunctionName( fd->name() );
```
It'd be a good idea to move sqlFunctionFromFunctionName and use of this up to the base QgsSqlExpressionCompiler class. This would make it easy to reuse for other providers.

rldhont replied:
@nyalldawson can you review my proposition about `sqlFunctionFromFunctionName` ? thanks

nyalldawson commented on src/providers/postgres/qgspostgresexpressioncompiler.cpp:
```diff
@@ -32,3 +36,91 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
+      if ( nd.isEmpty() )
+        return Fail;
+
+      result = QStringLiteral( "%1(%2)" ).arg( nd, args.join( QStringLiteral( "," ) ) );
```
.join(',') (the QChar overload) is MUCH faster.

mhugo reviewed (approved):
Looks good to me. Thanks !
As a bonus: Instead of inserts in the map, is there a way to have a const map filled during compilation ? This should be possible with QT5 + C++11 :
static const QMap<QString, QString> my_map = {{"intersects", "ST_intersects"}, {"...", "..."}};

And is there a way to add tests in the postgres provider tests ?

rldhont commented:
Thanks @mhugo, I'll update the way to defined map and add some tests.

To add test, I have to update `providertestbase.py` and all `test_provider_*.py`

mhugo commented:
@rldhont Adding something in test_provider_postgres.py is not enough ?

nyalldawson commented:
@mhugo no, it should be in providertestbase.py so all other providers can take also test the result and we know that there's consistent behavior with expression compilation and without. There's already a mature setup for these types of tests in providertestbase.py.

nyalldawson reviewed (changes requested):
I also mean that most of the compileNode changes here should be moved to QgsSqlExpressionCompiler. Specifically the argument node completion and all the code which relies on sqlFunctionFromFunctionName. Then it'll be easy to add function compilation to the other providers by just implementing sqlFunctionFromFunctionName in their compilers.

nyalldawson commented:
Also, to echo what @mhugo said, EVERY function which is being compiled here needs a test added in the provider test suite. It's critical that we keep consistent results between compiled/uncompiled expressions and across the different providers.

nyalldawson commented on src/providers/postgres/qgspostgresexpressioncompiler.cpp:
```diff
@@ -32,3 +36,161 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
+    fnNames.insert( QStringLiteral( "pi" ), QStringLiteral( "pi" ) );
+    // geometry functions
+    fnNames.insert( QStringLiteral( "azimuth" ), QStringLiteral( "ST_Azimuth" ) );
+    fnNames.insert( QStringLiteral( "x" ), QStringLiteral( "ST_X" ) );
```
We can't compile x/y as the QGIS versions have some differences. Eg qgis one will return centroid x for a polygon input.

nyalldawson commented on src/providers/postgres/qgspostgresexpressioncompiler.cpp:
```diff
@@ -32,3 +36,161 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
+    fnNames.insert( QStringLiteral( "overlaps" ), QStringLiteral( "ST_Overlaps" ) );
+    fnNames.insert( QStringLiteral( "within" ), QStringLiteral( "ST_Within" ) );
+    fnNames.insert( QStringLiteral( "translate" ), QStringLiteral( "ST_Translate" ) );
+    fnNames.insert( QStringLiteral( "buffer" ), QStringLiteral( "ST_Buffer" ) );
```
I don't think our buffer arguments match postgis' ordering/values.

rldhont commented:
@nyalldawson I have updated the way functions are compile. I have add a method `sqlAgrumentsFromFunctionName` and update the sql compiler. Are you agree with my proposition ?

@mhugo I will add the tests in an other commit.

nyalldawson commented on src/core/qgssqlexpressioncompiler.h:
```diff
@@ -93,6 +93,21 @@ class CORE_EXPORT QgsSqlExpressionCompiler
+     * @param fnName expression function name
+     * @returns the SQL function name
+     */
+    virtual QString sqlFunctionFromFunctionName( const QString& fnName );
```
Should be const

nyalldawson commented on src/core/qgssqlexpressioncompiler.h:
```diff
@@ -93,6 +93,21 @@ class CORE_EXPORT QgsSqlExpressionCompiler
+     * @param fnArgs arguments from expression
+     * @returns the arguments updated for SQL Function
+     */
+    virtual QStringList sqlArgumentsFromFunctionName( const QString& fnName, const QStringList& fnArgs );
```
Also const

nyalldawson commented on src/providers/postgres/qgspostgresexpressioncompiler.cpp:
```diff
@@ -32,3 +38,141 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
+        result = quotedIdentifier( mGeometryColumn );
+        return Complete;
+      }
+      else if ( fd->name() == "$area" )
```
$area/$length could be tricky, because the QGIS expression versions of these return ellipsoidal measurements based on the project settings, and also convert the result to the units specified in project properties. I'm not sure it's possible to compile these.

nyalldawson commented:
@rldhont great, that's exactly what I had in mind. I've been wanting to compile lower/upper for some time now add this will make it trivial. Thanks!

rldhont committed:
Sql Compiler: Add escape after IN operator

rldhont committed:
[Feature][PostgreSQL] Tests: Compile expression functions

rldhont commented:
@nyalldawson, @mhugo and @rouault I have some trouble with WFS provider tests.
2 expressions does not pass:
```
azimuth($geometry,geom_from_wkt( 'Point (-70 70)')) < pi()
touches($geometry,geom_from_wkt( 'Polygon ((-70.332 66.33, -65.32 66.33, -65.32 78.3, -70.332 78.3, -70.332 66.33))'))
```
For touches I think it's a problem with precision.
For azimuth I don't have any clue.

Are there any simple tests for expression ?

Thanks for your comments.

nyalldawson commented on tests/src/python/providertestbase.py:
```diff
@@ -240,6 +246,26 @@ def runGetFeatureTests(self, provider):
+        assert len([f for f in provider.getFeatures()]) == 4
+
+        # geometry
+        self.assert_query(provider, 'xmin($geometry) < -70', [1, 3])
```
Xmin/ etc also work with points in QGIS expressions

rldhont replied:
PostGIS autocast to box like in QGIS expressions

nyalldawson replied:
Ah yeah - what I meant (and should have said) was that these tests for xmin/etc should also be added for the point layer so that it's assured that if other (non postgis) providers compile xmin/etc that the compilation works for point layers.

rldhont replied:
I'll add it

nyalldawson commented on tests/src/python/providertestbase.py:
```diff
@@ -223,7 +223,13 @@ def runGetFeatureTests(self, provider):
+        self.assert_query(provider, 'y($geometry) > 70', [2, 4, 5])
+        self.assert_query(provider, 'disjoint($geometry,geom_from_wkt( \'Polygon ((-72.2 66.1, -65.2 66.1, -65.2 72.0, -72.2 72.0, -72.2 66.1))\'))', [4, 5])
         self.assert_query(provider, 'intersects($geometry,geom_from_wkt( \'Polygon ((-72.2 66.1, -65.2 66.1, -65.2 72.0, -72.2 72.0, -72.2 66.1))\'))', [1, 2])
+        self.assert_query(provider, 'touches($geometry,geom_from_wkt( \'Polygon ((-70.332 66.33, -65.32 66.33, -65.32 78.3, -70.332 78.3, -70.332 66.33))\'))', [1, 4])
```
Perhaps creating a simpler geometry using string concatenation could work. Something like "touches(geom_from_wkt('linestring(' || some attribute || ' 0, 1 0)'), ...)" Might help? I've only got phone access at the moment so can't make a more helpful expression sorry!

nyalldawson commented on src/providers/postgres/qgspostgresexpressioncompiler.cpp:
```diff
@@ -32,3 +38,150 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
+      { "pi", "pi" },
+      // geometry functions
+      { "azimuth", "ST_Azimuth" },
+      { "x", "ST_X" },
```
Don't forget that x/y work with non points in QGIS expressions too (they return centroid x/y). This will need to be handled explicitly eg by wrapping st_x(st_centroid(...

rldhont replied:
I'll do it

nyalldawson commented on src/providers/postgres/qgspostgresexpressioncompiler.cpp:
```diff
@@ -32,3 +38,150 @@ QString QgsPostgresExpressionCompiler::quotedValue( const QVariant& value, bool&
+        result = QStringLiteral( "ST_Perimeter(%1)" ).arg( quotedIdentifier( mGeometryColumn ) );
+        return Complete;
+      }
+      else if ( fd->name() == "$x" )
```
I think (from memory) $x/$y are an exact match for st_x/st_y

rldhont replied:
$x/$y does not exactly match st_x/st_y

nyalldawson commented:
I've posted a suggestion for touches, but not sure about azimuth. Maybe QGIS/postgis differ in the returned units (degree vs radians)?

rldhont commented:
Hi @nyalldawson, thanks for your comments.

About `azimuth`, what I don't understand is that the expression is tested in all other provider without compilation and works except with WFS.

About `touches`, I think it's an issue in WFS provider about coordinate precision reading.

About `x`and `$x', they are not the same method: `fcnGeomX` and `fcnX`:
```cpp
static QVariant fcnGeomX( const QVariantList& values, const QgsExpressionContext*, QgsExpression* parent )
{
  QgsGeometry geom = getGeometry( values.at( 0 ), parent );
  if ( geom.isEmpty() )
    return QVariant();

  //if single point, return the point's x coordinate
  if ( geom.type() == QgsWkbTypes::PointGeometry && !geom.isMultipart() )
  {
    return geom.asPoint().x();
  }

  //otherwise return centroid x
  QgsGeometry centroid = geom.centroid();
  QVariant result( centroid.asPoint().x() );
  return result;
}

static QVariant fcnX( const QVariantList&, const QgsExpressionContext* context, QgsExpression* )
{
  FEAT_FROM_CONTEXT( context, f );
  ENSURE_GEOM_TYPE( f, g, QgsWkbTypes::PointGeometry );
  if ( g.isMultipart() )
  {
    return g.asMultiPoint().at( 0 ).x();
  }
  else
  {
    return g.asPoint().x();
  }
}
```
It's why I do not compile `$x`,`$y`. But I have to update the `x` and `y` compilation.

Thanks again for your help.

nyalldawson commented:
I'm not sure why wfs would differ here. I can have a look when I'm back from holidays in ~2 weeks.

rldhont commented on src/core/qgssqlexpressioncompiler.cpp:
```diff
@@ -319,7 +319,7 @@ QgsSqlExpressionCompiler::Result QgsSqlExpressionCompiler::compileNode( const Qg
         return rn;
 
-      result = QStringLiteral( "%1 %2IN(%3)" ).arg( nd, n->isNotIn() ? "NOT " : "", list.join( QStringLiteral( "," ) ) );
+      result = QStringLiteral( "%1 %2 IN(%3)" ).arg( nd, n->isNotIn() ? "NOT " : "", list.join( QStringLiteral( "," ) ) );
```
The space has not been inserted at the right place. I'd like to add space between `IN` and `(`.

nyalldawson commented:
@rldhont I'm trying to work out what's left to address here. Does this sound like a fair summary:

Tests for
- sqrt
- radians
- degrees
- abs
- cos
- sin
- tan
- acos
- asin
- atan
- atan2
- exp
- ln
- log
- log10
- round
- floor
- ceil
- pi 
- test for x for poly
- test for y for poly
- z
- m
- x_min/x_max/etc for point
- translate (translate test needs to be removed from overlap test, because a provider may compile translate but not overlap which would mean this test isn't compiled and there's nothing checking translate)
- buffer (tests for segments/cap/join arguments) (also as above, needs splitting from overlaps test)
- split centroid, point_on_surface and intersects test (as above)
- reverse
- is_closed
- convex_hull
- difference
- distance
- intersection
- sym_difference
- combine
- union
- geom_from_gml

Also the failing WFS test

I'd suggest that to move things along you could remove the above functions and the failing touches test and then merge this PR. Then re-add the extra functions and tests in separate PRs.

This is great stuff and I'm keen to see it land, but I'd also prefer NOT to have any particular function compiled then to have it compiled but without a test. The bugs which could result from mis-compilation of a function would be very tricky to diagnose, which is why I'm being so picky to make sure every single one is covered.

rldhont commented:
Hi @nyalldawson,

I'll simplify my PR and merge it as you suggest.

I think we also needs complexe test. For exemple the expression `azimuth($geometry,geom_from_wkt( \'Point (-70 70)\')) < pi()` is compiled but `azimuth($geometry,geom_from_wkt( \'Point (-70 70)\')) < pi()/2` is not compiled.

Thanks for your reviews

rldhont commented:
Hi @nyalldawson,

I've updated tests, to cover:

- sqrt
- radians
- degrees
- abs
- cos
- sin
- tan
- acos
- asin
- atan
- atan2
- exp
- ln
- log
- log10
- round
- floor
- ceil
- pi
- test for x for poly
- test for y for poly
- x_min/x_max/etc for point
- geom_from_gml

About the functions that create geometries, I have no idea how to test them as a filter without intersects or overlaps function:

- translate
- buffer
- split centroid
- point_on_surface
- reverse
- convex_hull
- difference
- distance
- intersection
- sym_difference
- combine
- union
